R/peaks-functions.R

```diff
@@ -125,3 +127,66 @@ NULL
     x <- x[x[, 2] > .qtl, 1]
     quantile(diff(x), 0.25) > k
 }
+
+#' @description Compare peaks between 2 spectra
+#'
+#' @description
+#'
+#' Compare peaks from one spectrum to peaks from another spectrum with the
+#' provided function (`FUN`). `.peaks_compare_intensities` ensures that peaks
+#' are first matched between the two spectra with the [closest()] function.
+#'
+#' @param x peaks `matrix` with columns `"mz"` and `"intensity"`.
+#'
+#' @param y peaks `matrix` with columns `"mz"` and `"intensity"`.
+#'
+#' @param FUN `function` to be applied to the intensity values of matched
+#'     peaks between peak lists `x` and `y`. The function should take two
+#'     `numeric` vectors as their first argument and should return a single
+#'     `numeric`. Parameter `...` will be passed to the function.
+#'
+#' @param tolerance `numeric(1)` with the acceptable (constant) difference in
+#'     m/z values for matching peaks.
+#'
+#' @param ppm `numeric(1)` specifying a peak-specific, relative,
+#'     tolerance for the peak matching. Peaks between the two spectra with a
+#'     difference in their m/z smaller than *ppm* of the m/z of the peak (of
+#'     the second spectrum) are matched. See also [ppm()].
+#'
+#' @param ... additional parameters passed to the `FUN` function.
+#'
+#' @return The result from `FUN`, which should be a `numeric(1)`.
+#'
+#' @author Johannes Rainer
+#'
+#' @noRd
+#'
+#' @importFrom stats cor
+#'
+#' @importFrom MsCoreUtils closest ppm
+#'
+#' @examples
+#'
+#' x <- cbind(c(31.34, 34.43, 54.65, 300.12, 514.5, 856.12),
+#'      c(13, 15, 56, 7, 45, 321))
+#'
+#' y <- cbind(c(34.431, 35.34, 50.24, 300.121, 514.51, 856.122),
+#'      c(123, 34, 323, 432, 12, 433))
+#'
+#' Spectra:::.peaks_compare_intensities(x, y)
+#' Spectra:::.peaks_compare_intensities(x, y, ppm = 40)
+#' Spectra:::.peaks_compare_intensities(x, y, ppm = 0)
+#'
+#' ## To get the number of common peaks
+#' Spectra:::.peaks_compare_intensities(x, y, FUN = function(x, y) length(x))
+#' Spectra:::.peaks_compare_intensities(x, y, FUN = function(x, y) length(x), ppm = 40)
+#' ## NA if there are none common
+#' Spectra:::.peaks_compare_intensities(x, y, FUN = function(x, y) length(x), ppm = 0)
+.peaks_compare_intensities <- function(x, y, FUN = cor, tolerance = 0,
+                                       ppm = 20, ...) {
+    matches <- closest(x[, 1], y[, 1], tolerance = tolerance + ppm(y[, 1], ppm))
```

Comment:
Shouldn't we use `closest(x[, 1], y[, 1], ... , duplicates="closest")` here to avoid duplicated matches to `y`?

Reply:
Yes, makes sense. Will change that.